File: src/main/java/dev/kikugie/techutils/feature/worldedit/WorldEditSync.java
package dev.kikugie.techutils.feature.worldedit;

import com.mojang.brigadier.CommandDispatcher;
import dev.kikugie.techutils.TechUtilsMod;
import dev.kikugie.techutils.config.WorldEditConfigs;
import dev.kikugie.techutils.util.ResponseMuffler;
import dev.kikugie.techutils.util.ValidBox;
import fi.dy.masa.litematica.data.DataManager;
import net.minecraft.client.MinecraftClient;
import net.minecraft.command.CommandSource;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Objects;
import java.util.Optional;

public class WorldEditSync {
    @Nullable
    private static WorldEditSync instance = null;
    private final MinecraftClient client = MinecraftClient.getInstance();
    private final WorldEditNetworkHandler handler = WorldEditNetworkHandler.initHandler();
    private ValidBox lastBox;
    private int syncTick = 0;
    private boolean worldEditConnected = false;
    private boolean commandAvailable = false;
    private boolean initComplete = false;

    public static void init() {
        instance = new WorldEditSync();
    }

    public static Optional<WorldEditSync> getInstance() {
        return Optional.ofNullable(instance);
    }

    public void onTick() {
        boolean initCompleteCache = !this.initComplete && this.worldEditConnected && this.commandAvailable;
        this.initComplete = this.worldEditConnected && this.commandAvailable;
        if (initCompleteCache)
            disableNeighborUpdates();
        if (this.initComplete)
            syncSelection();
    }

    public void onWorldEditConnected() {
        this.worldEditConnected = true;
    }

    public void onCommandTreePacket(CommandDispatcher<CommandSource> dispatcher) {
        this.commandAvailable = dispatcher.findNode(List.of("/pos1")) != null;
    }

    private void disableNeighborUpdates() {
        if (!WorldEditConfigs.DISABLE_UPDATES.getBooleanValue())
            return;
        ResponseMuffler.scheduleMute("Side effect \"Neighbors\".+");
        Objects.requireNonNull(this.client.getNetworkHandler()).sendCommand("/perf neighbors off");
        TechUtilsMod.LOGGER.debug("Turning off WorldEdit neighbor updates");
    }

    private void syncSelection() {
        if (!WorldEditConfigs.WE_SYNC.getBooleanValue())
            return;

        if (this.syncTick > 0) this.syncTick--;
        if (!this.handler.storage.isCuboid())
            return;

        Optional<ValidBox> boxOptional = getActiveSelection();
        if (boxOptional.isEmpty())
            return;
        ValidBox box = boxOptional.get();

        if (!box.equals(this.lastBox)) {
            this.lastBox = box;
            this.syncTick = WorldEditConfigs.WE_SYNC_TICKS.getIntegerValue();
            return;
        }
        if (this.syncTick != 0)
            return;

        this.syncTick = -1;
        updateRegion(box);
        TechUtilsMod.LOGGER.debug("WorldEdit synced!");
        if (WorldEditConfigs.WE_SYNC_FEEDBACK.getBooleanValue())
            this.client.player.sendMessage(Text.translatable("techutils.feature.worldeditsync.success").formatted(Formatting.GREEN), true);
    }

    private void updateRegion(ValidBox box) {
        ResponseMuffler.scheduleMute("(\\w+ position set to \\(.+\\).)|(Position already set.)");
        this.client.getNetworkHandler().sendCommand(String.format("/pos1 %d,%d,%d",
                box.getPos1().getX(),
                box.getPos1().getY(),
                box.getPos1().getZ()));
        ResponseMuffler.scheduleMute("(\\w+ position set to \\(.+\\).)|(Position already set.)");
        this.client.getNetworkHandler().sendCommand(String.format("/pos2 %d,%d,%d",
                box.getPos2().getX(),
                box.getPos2().getY(),
                box.getPos2().getZ()));
    }

    private Optional<ValidBox> getActiveSelection() {
        var selection = DataManager.getSelectionManager().getCurrentSelection();
        if (selection == null) {
            return Optional.empty();
        }
        var box = selection.getSelectedSubRegionBox();
        if (box == null || box.getPos1() == null || box.getPos2() == null) {
            return Optional.empty();
        }
        return Optional.of(new ValidBox(box.getPos1(), box.getPos2(), "region"));
    }
}

File: src/client/kotlin/dev/kikugie/techutils/impl/worldedit/WorldEditSync.kt
package dev.kikugie.techutils.impl.worldedit

import dev.kikugie.techutils.config.WorldEditConfig
import dev.kikugie.techutils.util.InGameNotifier
import dev.kikugie.techutils.util.PlacementUtils
import fi.dy.masa.litematica.selection.Box
import fi.dy.masa.malilib.gui.Message
import net.minecraft.client.MinecraftClient
import net.minecraft.util.math.BlockBox

object WorldEditSync {
    val client = MinecraftClient.getInstance()
    val handler
        get() = WorldEditNetworkHandler.INSTANCE
    val queue = MuteQueue()
    var tick = 0
    var lastBox: BlockBox? = null

    private val REGEX1 = Regex("Side effect \"Neighbors\".+")
    private val REGEX2 = Regex("(\\w+ position set to \\(.+\\).)|(Position already set.)")

    fun onConnected() {
        if (WorldEditConfig.disableUpdates.booleanValue) disableNeighborUpdates()
    }

    fun tick() {
        if (handler?.shouldTick == true && WorldEditConfig.sync.booleanValue) syncSelection()
    }

    private fun disableNeighborUpdates() {
        queue += REGEX1
        client.networkHandler?.sendCommand("/perf neighbors off")
    }

    private fun syncSelection() {
        if (tick > 0) tick--
        if (handler?.storage?.cuboid != true) return
        val box = PlacementUtils.selection?.selectedSubRegionBox?.toIntBox()
            ?: return
        if (box != lastBox) {
            lastBox = box
            tick = WorldEditConfig.syncTicks.integerValue
            return
        }

        if (tick != 0) return
        tick = -1
        updateRegion(box)
        if (WorldEditConfig.syncFeedback.booleanValue)
            InGameNotifier.addMessage(Message.MessageType.SUCCESS, "techutils.worldedit_sync.synced")
    }

    private fun updateRegion(box: BlockBox)  {
        queue += REGEX2
        client.networkHandler?.sendCommand("/pos1 ${box.minX},${box.minY},${box.minZ}")
        queue += REGEX2
        client.networkHandler?.sendCommand("/pos2 ${box.maxX},${box.maxY},${box.maxZ}")
    }
}

private val Box.isValid
    get() = pos1 != null && pos2 != null

private fun Box.toIntBox() = if (isValid) BlockBox.create(pos1, pos2) else null